### ISSUE TITLE
Add PostgreSql Support

### ISSUE THREAD

**Issue by SimonCropp** [Particular/NServiceBus.Persistence.Sql#15]

**Comment by weralabaj** [comment]
Consider raising it when we implement AWS support and/or once we have a proper .NET core support.

**Comment by janovesk** [comment]
This would also be nice for Maarten users, btw.

**Comment by asiemer** [comment]
I have a large client that would love to be able to use PostgreSQL (migrating away from Raven) hosted on RDS in AWS.  Thank you.

**Comment by weralabaj** [comment]
Hi @asiemer, since we're talking about moving away Raven, are they interested in Postgres or Marten in particular?

**Comment by asiemer** [comment]
All the spikes I am currently doing are around Marten and Postgres.  Given our timeline I am suggesting a migration to MSSQL/RDS for NSB and app data through Marten/PostgreSQL/RDS.  They like the ease of development and have a significant investment in patterns around Raven.  Need an easy way to swap out Raven and in something else.  Marten/Postgresql solves this it seems on the app dev side.  Once you all release ado/postgres support for NSB...that can follow.  Thoughts?

**Comment by asiemer** [comment]
Is there perhaps an early adopter release of this Postgresql persistence mechanism?  I would love to see how it swaps into their current set up.  Give some feedback.

**Comment by weralabaj** [comment]
@asiemer no, we don't have anything ready at the moment, but knowing there's an interest can affect the priority we give to that idea.

**Comment by masterpoi** [comment]
We are also migrating away from Raven to a Postgresql/Marten setup. Support for postgres in NSB would provide awesome complimentary for us.

**Comment by weralabaj** [comment]
@asiemer @masterpoi We're working on supporting PostgreSQL right now. Here's the PR https://github.com/Particular/NServiceBus.Persistence.Sql/pull/183.

I'm not sure if it's still relevant for you, but if you'd be interested in providing feedback and/or testing it once it's ready, let me know and we'll arrange that.

For now we don't plan on providing special support for Marten though, just standard PostgreSQL.

**Comment by weralabaj** [comment]
@asiemer @masterpoi We're getting close to getting the full Postgres implementation. Would any of you be interested in testing it or discussing in more detail?

Having an external feedback is always appreciated, and can help us catch aspects we didn't consider. If interested, you can let us know here or email me at weronika.labaj at particular.net.

**Comment by masterpoi** [comment]
@weralabaj Sure. We can set it up in one of our new projects and share our experiences.  Mind, we are on .NET core, so using NServiceBus 7 beta. We currently have no easy way to try it in 6.0 / full framework.

**Comment by weralabaj** [comment]
That will be released on top of v7, so it's really perfect! :)

**Comment by masterpoi** [comment]
@weralabaj I did some short tests in one of our .NET core projects replacing InMemoryPersistence with SqlPersistence. (v3.1.0-pullrequest0183-0066)  I didn't have any mayor or minor issues really, and everything seems to be working right. 
Things that were new to me, coming from RavenDb:

- Had to create the database manually.
- Had to add the JsonBParameterModifier thing, which was clearly indicated by the exception message.
- Saga -> SqlSaga
- Need to incorporate the sql scripts in our build/deployment plans and deal with versioning.

I've tested some sagas and the outbox functionality and everything behaved as expected. 
With rabbitmq i noticed the TimeoutData table to be created, which i did not expect since, if i understood correctly, rabbitmq manages it's own timeouts. In any case, delayed delivery or saga timeouts do not create records in the table. When using learning transport the table is not created.

We will continue to use the library during our developments for now and let you know if we run into any issues.

**Comment by masterpoi** [comment]
Adding a 

```csharp 
[assembly: SqlPersistenceSettings(MsSqlServerScripts = false, MySqlScripts = false, OracleScripts = false, PostgreSqlScripts = true)]
```
results in a  

> SqlPersistenceTask: Must define at least one of MsSqlServerScripts, MySqlScripts, or OracleScripts. Add a [SqlPersistenceSettingsAttribute] to the assembly. error during building

**Comment by SimonCropp** [comment]
@masterpoi 

> SqlPersistenceTask: Must define at least one of MsSqlServerScripts, MySqlScripts, or OracleScripts. Add a [SqlPersistenceSettingsAttribute] to the assembly. error during building

that should be fixed if you update to `3.1.0-pullrequest0203-0075`

**Comment by SimonCropp** [comment]
sorry that should be `3.1.0-pullrequest0183-0094`

**Comment by weralabaj** [comment]
Thanks @masterpoi !  I'll look into the aspects you highlighted, it's very helpful :)

**Comment by weralabaj** [comment]
@masterpoi Regarding your timeout table question, I've verified it manually, the table is created by default indeed, but if you explicitly disable TimeoutManager then it' won't be created.
This is for backwards compatibility in Rabbit [details](https://docs.particular.net/transports/rabbitmq/delayed-delivery#backwards-compatibility), i.e. by default Rabbit still supports endpoints running on older versions that don't have native timeout delivery (that was only introduced in NServiceBus.RabbitMQ v 4). If you are sure that all your endpoints run on NServiceBus.RabbitMQ v4 or above, then you can explicitly disable the TimeoutManager and rely on native timeouts only. Then the table won't be created.

**Comment by weralabaj** [comment]
BTW do you plan on migrating any system from RavenDB persistence to Postgres? If so, we have [a guide](https://github.com/ParticularLabs/RavenToSqlPersistence) that may help and would be great to know in case you need any assistance :)

**Comment by masterpoi** [comment]
@weralabaj Oh, ic, that makes sense. We are not running any older RabbitMQ endpoints, so it's good to know we can disable the TimeoutManager. Thanks for the link, i hadn't seen that document yet. Also, that native delay mechanic look really fancy :)

With regard to RavenDB -> Postgres, fortunately we do not need to migrate the Saga data, since our Sagas are short lived, and have relatively low traffic. This makes a cut-over migration possible. Thanks for the information though!

**Comment by masterpoi** [comment]
Just fyi, the following is something subtle i ran into. It is not a bug at all, but probably worth mentioning in the docs. (Or maybe logging a warning or something)

We have a saga that has a property containing a polymorphic collection. To serialize it you do the obvious and add 
```csharp
           persistence.SagaSettings().JsonSettings(new JsonSerializerSettings
            {
                TypeNameHandling = TypeNameHandling.Auto,
            });
```
to the startup configuration. Newtonsoft.Json adds a $type metadata property to the members of the collection to indicate their type. By default when deserializing it expects this property to be the 
first property of the object. However, *Postgresql jsonb columns might change the order of how properties are stored*. The $type property might not be the first property when the data comes back from the database. Hence you need to add 
```   
MetadataPropertyHandling = MetadataPropertyHandling.ReadAhead 
```
to your startup config as well so that the deserializer reads the whole object before looking for the $type property.

**Comment by SimonCropp** [comment]
@masterpoi 

that is very useful information. thanks

> Postgresql jsonb columns might change the order of how properties are stored

is there any doco link for that behaviour?

**Comment by SimonCropp** [comment]
@masterpoi ok i found the relevant link and added a section in our doco about it https://docs.particular.net/persistence/sql/?version=sqlpersistence_3.1#usage-postgresql-json-net-typenamehandling

**Comment by masterpoi** [comment]
@SimonCropp Perfect! Thanks.

**Comment by weralabaj** [comment]
Done, released in 3.0 and 4.0.
